### DISCUSSION TITLE
How exactly do lifecycles work?

### DISCUSSION THREAD

**Discussion by ribru17** [solidjs/solid#1819] [Q&A]
I have some code in my `root.tsx` file that looks like this:
```tsx
onMount(() => {
  inject();
  if (!isServer) {
    document.addEventListener('mousedown', handleClickOutside);
  }
});
```
If I remove the server check, the website breaks because `document` does not exist. Same error occurs with the cleanup code in `onCleanup`. I was confused because I read in the documentation online that lifecycles like `onMount` only run in the browser, and thus you wouldn't need to check if you are in the server or not. I did notice, however, that any lifecycle components that were not in my root file (e.g. in some other component pulled in later) these lifecycles would correctly run only in the browser, no server checking necessary. Is this only the case for the root file or am I missing something?

**Comment by snnsnn** [comment]
Solid components do not have an actual life cycle. `onMount` creates an effect internally and wraps in inside `untrack` so that the effect is run once. Effects are run after the rendering phase, so onMounts creates an effect that is run once when component renders.

`onCleanup` function pushes its callback to its owner, that is a computation in which `onCleanup` is called. When this computation is disposed, Solid's runtime calls the previously stored callbacks.

Here is how `onMount` is defined on the server side:

```tsx
export function onMount(fn: () => void) {}
```

See, it is a void function. So, the callback should never run.

> If I remove the server check, the website breaks because document does not exist.

If you are sure it is the `onMount` call, you can refactor your conditional logic in way to check if `isServer` before calling `onMount`.
